dbms/src/AggregateFunctions/AggregateFunctionMLMethod.h

```diff
@@ -0,0 +1,551 @@
+#pragma once
+
+#include <type_traits>
+
+#include <IO/WriteHelpers.h>
+#include <IO/ReadHelpers.h>
+
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/DataTypesDecimal.h>
+
+#include <Common/FieldVisitors.h>
+#include <Columns/ColumnVector.h>
+#include <Columns/ColumnsCommon.h>
+#include <Columns/ColumnsNumber.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <Functions/FunctionHelpers.h>
+
+#include <cmath>
+#include <exception>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+}
+
+/**
+GradientComputer class computes gradient according to its loss function
+*/
+class IGradientComputer
+{
+public:
+    IGradientComputer()
+    {}
+
+    virtual ~IGradientComputer() = default;
+
+    /// Adds computed gradient in new point (weights, bias) to batch_gradient
+    virtual void compute(std::vector<Float64> * batch_gradient, const std::vector<Float64> &weights, Float64 bias,
+                         Float64 learning_rate, Float64 l2_reg_coef, Float64 target, const IColumn **columns, size_t row_num) = 0;
+
+    /// Now we should use predict_block function instead of predict
+    virtual void predict_block(ColumnVector<Float64>::Container &container,
+                                 Block &block, const ColumnNumbers &arguments,
+                                 const std::vector<Float64> &weights,
+                                 Float64 bias) const = 0;
+};
+
+
+class LinearRegression : public IGradientComputer
+{
+public:
+    LinearRegression()
+    {}
+
+    void compute(std::vector<Float64> * batch_gradient, const std::vector<Float64> &weights, Float64 bias,
+                 Float64 learning_rate, Float64 l2_reg_coef, Float64 target, const IColumn **columns, size_t row_num) override
+    {
+        Float64 derivative = (target - bias);
+        for (size_t i = 0; i < weights.size(); ++i)
+        {
+            derivative -= weights[i] * static_cast<const ColumnVector<Float64> &>(*columns[i]).getData()[row_num];
+        }
+        derivative *= (2 * learning_rate);
+
+        (*batch_gradient)[weights.size()] += derivative;
+        for (size_t i = 0; i < weights.size(); ++i)
+        {
+            (*batch_gradient)[i] +=
+                    derivative * static_cast<const ColumnVector<Float64> &>(*columns[i]).getData()[row_num]
+                        - 2 * l2_reg_coef * weights[i];
+        }
+    }
+
+    void predict_block(ColumnVector<Float64>::Container &container,
+                         Block &block,
+                         const ColumnNumbers &arguments,
+                         const std::vector<Float64> &weights, Float64 bias) const override
+    {
```

Comment:
Check `weights.size() + 1== arguments.size()`

Reply:
This condition is checked in predictResultInto (line ~520)

Reply:
This condition is checked in `predictValue()` (line ~520)

Reply:
Anyway it's better to check in case you change invariant and forget to fix some code.